integation of skydb v1 and skydb v2 + test script

parajbs opened the pull request:
changelog:
- Integration von skydb v1
- Integration von skydb v2
- create a test script in "scripts/skydb.js"

# PULL REQUEST

## Overview

## Example for Visual Changes
<!--
For user facing features please provide proof that the format is as expected.
Screen shots and/or asciinema recordings are very helpful.
-->

## Checklist
Review and complete the checklist to ensure that the PR is complete before assigned to an approver.
 - [ ] All new methods or updated methods have clear docstrings
 - [ ] Testing added or updated for new methods
 - [ ] Verify if any changes impact the WebPortal Health Checks
 - [ ] Approriate documentation updated
 - [ ] Changelog file created

## Issues Closed
<!--
Use the `Closes` keyword to automatically close the issue on merge.
Example: Closes #XXXX
-->

parajbs commented:
Hello developer

i need help again I'm not very familiar with Github.
Can someone explain to me how I can get rid of the errors?
I don't understand the mistakes.
With the error:

**kodiakhq: status - Ignored (no automerge label: 'automerge')**

I had it before and thought that the error came because when I did a PR on a branch.
But this time I did the PR on the master branch and the error is back. Help.

**And why are the tests not passing?**

**And what does the error with write access mean?**

**How can I shoot out the errors before I send the PR?**

Would be very grateful for help and short explanation.

m-cat commented:
Just glancing at it, looks great! I'll give it a good review when I have time.

Can this be closed? https://github.com/SkynetLabs/nodejs-skynet/pull/140

m-cat commented:
> Hello developer
> 
> i need help again I'm not very familiar with Github. Can someone explain to me how I can get rid of the errors? I don't understand the mistakes. With the error:
> 
> **kodiakhq: status - Ignored (no automerge label: 'automerge')**
> 
> I had it before and thought that the error came because when I did a PR on a branch. But this time I did the PR on the master branch and the error is back. Help.
> 
> **And why are the tests not passing?**
> 
> **And what does the error with write access mean?**
> 
> **How can I shoot out the errors before I send the PR?**
> 
> Would be very grateful for help and short explanation.

I had to approve the tests as you're a third-party contributor. 

The Kodiak job only runs for Dependabot PRs so don't worry about that one.

You can look at the details for each job to see what the error(s) are.

> How can I shoot out the errors before I send the PR?

You have to install Yarn and then run `yarn` followed by `yarn lint` (to run code quality checks) and `yarn test` (to run tests).

parajbs commented:
thanks for the explanation and guidance.
I will try it next time.

I looked at the 5 mistakes.
I was able to fix the first and second error.
Now there are still 3 errors with "BigInit":

error 'BigInt' is not defined no-undef

Not sure about this error as I have the code from "skynet-js".
Since it was running I didn't think about it.
I'll have to google it and do some tests.
If you have an idea?

parajbs commented:
hello m-cut

I tested yarn and yarn lint.
The errors with "BigInt" come up there too.
I found something on google and tested it, and it works.

I had to make a change in the ".eslintrc.json" file under "env" area because "BigInt" is only supported from "es2020" onwards:
    "es6": true,
to
    "es2020": true,"

Is it a problem to change "es6" to "es2020" to support "BigInt"?

Do I have to create a new PR and delete this one, or can I somehow make direct changes here?

parajbs commented:
I still have to create a file "src/skydb.test.js" because I now know where it is needed for.
Coming soon, I have to see if I can understand it and get it right.

Thanks for the information.
I need to change my developer environment a bit.

m-cat commented:
> hello m-cut
> 
> I tested yarn and yarn lint. The errors with "BigInt" come up there too. I found something on google and tested it, and it works.
> 
> I had to make a change in the ".eslintrc.json" file under "env" area because "BigInt" is only supported from "es2020" onwards: "es6": true, to "es2020": true,"
> 
> Is it a problem to change "es6" to "es2020" to support "BigInt"?
> 
> Do I have to create a new PR and delete this one, or can I somehow make direct changes here?

Please don't make a new PR! You can update this one by making a new commit and pushing to your branch.

It's totally fine to upgrade to `es2020`. I believe we have that in `skynet-js` so please look at the configuration there and copy over any relevant parts. If there's any build issues they will be caught by our CI ("Continuous Integration", builds and tests your code automatically).

m-cat commented:
Please make sure you've read our [Contributing guideline](https://github.com/SkynetLabs/.github/blob/master/CONTRIBUTING.md) to make sure you understand the workflow, it will make it a lot easier for you. `git`, GitHub, yarn, CI, etc. are all standard these days so understanding them goes a long way in the Open Source world.

parajbs commented:
Hello m-cat

I made all the changes and then made a "push".
I also learned a lot and have now changed my developer environment.
all tests were OK.

Since I'm not very familiar with Github yet, I'll keep my fingers crossed that all Gihhub tests will run without errors and that it will continue.

Sorry again for my difficulties with the code bugs, but I didn't know the tools "prettier" / "lint" and jest.
I used github as a bookmark for interesting projects before.

m-cat commented on scripts/skydb.js:
```diff
@@ -0,0 +1,95 @@
+/**
+ * Demo script that uploads data passed in as CLI arguments.
+ *
+ * Example usage: node scripts/skydb.js <dbVersion> <dataKey>
```
It would be cool if you could also specify the data as a command line argument.

m-cat replied:
It would be cool if you could also specify the data as a command line argument.

m-cat commented on scripts/skydb.js:
```diff
@@ -0,0 +1,95 @@
+const process = require("process");
+
+const { SkynetClient, genKeyPairFromSeed } = require("..");
+const { publicKey, privateKey } = genKeyPairFromSeed("Super Save Test Secret");
```
I think it would be cool to print out the keys, that way the user can get to his data again. Also cool if you could (optionally) specify the private key.

m-cat commented on scripts/skydb.js:
```diff
@@ -0,0 +1,95 @@
+ *
+ * Example for Skydb v1 usage: node scripts/skydb.js -v1 testKey
+ *
+ * Example for Skydb v2 usage: node scripts/skydb.js -v2 testKey
```
Maybe the version could be optional and default to v2.

m-cat replied:
Maybe the version could be optional and default to v2.

m-cat commented:
Looks awesome! I only tried the script so far and will try to do a full review soon.

Thank you so much!

m-cat commented:
Looks awesome! I only tried the script so far and will try to do a full review soon.

Thank you so much!

parajbs commented:
Hello m-cut

**What is the script "script/skydb.js" supposed to be for?
I think it would be a test script to see if all functions are working. Or?**

Namely, I have already written a new, much more complex version of it, where all the integrated functions of skydb v1 and v2 are listed one after the other in a specific order with fixed data.
All functions work great!
You can (**select the versions / set the dataKey and set the seed**) as parameters.
**I also have the publicKey and privateKey in the console.log output.**

**I can write the individual versions in two files (skydb_v1.js and skydb_v2.js).
am I supposed to do this?**

Data also as a command line argument is **currently not possible with the new version.**

Since I programmed the new script in a much more complex way, the script only works with the defaut data, which is fixed so that all other functions also run through.
**In the next version maybe if I can do it.**

I clean the code from my new version and "push" it up.
this includes "scripts/skydb_v1.js" and "scripts/skydb_v2.js".
 
Then I do another "push" with the next version (include "scr/skydb.test.js).
But here it can lead to problems, since all skydb functions are integrated and tested there as well.
The problem is that the test takes a little longer than 5 seconds.

It took me some time to get it local in that it runs all tests cleanly. And this several times in a row without breaking off the test because the time has expired.
Hope it works on github too.

parajbs commented:
I looked at the code and I can also enable data as a command line argument.
**But it must be specified like below no 1:**
**1.**
{\\"example\\":\\""This is some example JSON data\\"}

**Not like this**:
2.
{ "example": "This is some example JSON data" }

**Also not**:
3.
{ example: "This is some example JSON data" }

**Neither**:
4.
'{ "example":"This is some example JSON data" }'

**ONLY 1. is OK!**

Open to suggestions.

parajbs commented:
hello m-cat

I have also created the file "scr/skydb.test.js" and hope that it will also work on github.
It works fine locally.

Just saw that it was accepted as ok on github. Cool!!!

Now I have to wait for you, I think all functions should work.
Since I didn't find any examples for some functions and nothing in the documentation either, I had to create them myself.
I hope that I understood everything correctly to create examples.
Just take a look. Please give feedback.
Thanks in advance.

parajbs commented:
To all **developers**,

I have a question:
Should I test other functions of "skydb-js" and try to integrate?

**These functions are available for selection:**

a. from the **"download area"**:
a1. getFileContent
a2. getFileContentBinary
a3. openFile
**important !!! I can't test the other ones with hns at the moment because I don't have an HNS domain to test yet! But can see through them if they should work, and you have to test them and let me know.**
a4. downloadFileHns
a5. getHnsUrl
a6. getHnsresUrl
a7. getFileContentHns
a8. getFileContentBinaryHns
a9. openFileHns
a10. resolveHns

b. from the **"pin area"**:
b1. pinSkylink

c. from the **"MySky area"**:
**I tried to integrate this before, but I had problems. But I can try again. Complicated.**
c1. extractDomain
c2. getFullDomainUrl
c3. loadMySky

d. from the **"File area"**:
d1. getJSON
d2. getJSONEncrypted

**Please let me know.**
Additionally, I'm still working on a new feature, as a surprise.

parajbs commented:
hello m-cat,

I just have a somewhat changed version that is for more compatibility and orginalism to "skynet-js".
The code now looks more similar to "skynet-js".

and I have changed the scripts so that you can use a Jsondata file as a data source.

have fun with it.
tonight I also have additionally a new feature.

m-cat commented:
Sorry for the delay, getting to this now. I just saw all the things you added - awesome! Thank you for going above and beyond with this PR! Reviewing now.

m-cat commented on src/skydb.test.js:
```diff
@@ -0,0 +1,154 @@
+  axios.mockResolvedValue({ data: { skylink } });
+});
+
+describe("db.setDataLink", () => {
+  it("should be a DataLink set", async () => {
+    await client.db.setDataLink(privateKey, dataKey, dataLink);
+  });
+});
+
+describe("db.setEntryData", () => {
+  it("should set EntryData", async () => {
+    await client.db.setEntryData(privateKey, dataKey, rawEntryData);
+  });
+});
```
You can use `describe` to create groups, e.g.:

```suggestion
describe("SkyDB V1", () => {
  describe("setDataLink", () => {
    it("should be a DataLink set", async () => {
      await client.db.setDataLink(privateKey, dataKey, dataLink);
    });
  });
  
  describe("setEntryData", () => {
    it("should set EntryData", async () => {
      await client.db.setEntryData(privateKey, dataKey, rawEntryData);
    });
  });

...
}
```

Same for V2.

parajbs replied:
I'm not sure if I got it right, but I'm trying my best.

But I have a question. How do you paste these code images here in gibhub? 
Would be helpful for me, but I didn't do it myself.

m-cat replied:
You can either use [markdown](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#code) or you can insert a "suggestion" (by clicking one of the buttons in the comment toolbar.)

m-cat commented on src/skydb.test.js:
```diff
@@ -0,0 +1,154 @@
+});
+
+describe("db.deleteJSON", () => {
+  it("should delete jsonData on skydb", async () => {
```
Please add a test that verifies that the `jsonData` has been deleted.

parajbs replied:
Yes, I just saw it. 
I add one more "db.getJSON" with "await expect(receivedData.data).toEqual(null);" for confirming that verifies that the jsonData has been deleted.

m-cat commented on src/skydb.test.js:
```diff
@@ -0,0 +1,154 @@
+describe("dbV2.getRawBytes", () => {
+  it("should get RawBytes", async () => {
+    const receivedData = await client.dbV2.getRawBytes(publicKey, dataKey);
+    await expect(receivedData["dataLink"]).toEqual(dataLink || dataLink_dbV2);
```
I don't think this does what you intend it to do. The `||` will make this evaluate to `dataLink`. Try opening `node` in your terminal and doing this:

```
> 3 || "asdf"
3
```

Is there a way to change this test to only expect one value?

parajbs replied:
I'll have to look at what I've done there and improve it.
Since "yarn test" and "yarn lint" ran clean, I thought everything was OK and would work.
And I didn't think about it because of it.

m-cat replied:
No worries, it's common for small things like this to escape notice. That's why we do code reviews. 🙂

m-cat commented on src/skydb.test.js:
```diff
@@ -0,0 +1,154 @@
+  });
+});
+
+describe("dbV2.setJSON", () => {
+  it("This is important for the next tests with jest", async () => {
+    const receivedData = await client.db.setJSON(privateKey, dataKey, data);
+
+    await expect(receivedData.dataLink).toEqual(`sia://${skylink}`);
+  });
+});
```
I don't understand what this test is for.

parajbs replied:
Since all tests from skdb v1 and v2 are in one file and run one after the other, and the same variables are used for skydb v1 and v2, I had the problem when creating the test from dbV2 that it didn't work without existing jsonData.

That's why I set new data on Skydb with db.setJSON before the test dbV2 starts.

I can try again to leave this out. But I can not promise anything!
I think I can run it as a final test of skydb v1 but it would have the same effect.

m-cat replied:
I would use different data for V1 and V2, if possible, or even random data. But if that's too much work you can just leave it as-is - I know you put a lot of time into this code, already.

m-cat commented on src/defaults.js:
```diff
@@ -0,0 +1,102 @@
+  portalFileFieldname: "file",
+};
+
+const json_response_version = 2;
```
I really like the idea of a `defaults.js` file! However, all the SkyDB constants and helper methods don't really fit here.

m-cat replied:
The standard convention is that constants should be all uppercase:

```suggestion
const JSON_RESPONSE_VERSION = 2;
```

parajbs replied:
In my new version on my dev server, I already built it back.
I was tempted by the "defaultOptions" variable to write all things that I ported from "skynet-js" in lower case.

But the next thing I did was write everything back to the original for compatibility and understanding.

But then I got the idea, why porting everything, if you can change a few lines in "package.json" from "SkynetLabs/skynet-js", also "skynet-js" can bring all variables and functions for the import to be made available in "SkynetLabs/nodejs-skynet".
I think it's simpler and more complicated than porting.

look here:
https://github.com/SkynetLabs/skynet-js/issues/468

m-cat commented on src/defaults.js:
```diff
@@ -0,0 +1,102 @@
+ * @param data - The given JSON data.
+ * @returns - The Skynet JSON data.
+ */
+const buildSkynetJsonObject = async function (data) {
```
```suggestion
const buildSkynetJsonObject = function (data) {
```

parajbs replied:
can i change
but maybe later we can import the function directly from "skynet-js". 
Let's see if my request goes through with "skynet-js".

look here:
https://github.com/SkynetLabs/skynet-js/issues/468

m-cat commented on src/defaults.js:
```diff
@@ -0,0 +1,102 @@
+const defaultGetEntryOptions = {
+  ...defaultBaseOptions,
+  endpointGetEntry: "/skynet/registry",
+  hashedDataKeyHex: false,
```
This field should be removed since it appears unused. Same below in `defaultSetEntryOptions`.

parajbs replied:
In the new version it is used, but with the original name of "skydb-js".

DEFAULT_GET_ENTRY_OPTIONS

Or did they mean something else?

m-cat commented on src/defaults.js:
```diff
@@ -0,0 +1,102 @@
+};
+
+const defaultGetEntryOptions = {
+  ...defaultBaseOptions,
+  endpointGetEntry: "/skynet/registry",
```
```suggestion
  ...defaultOptions("/skynet/registry"),
```

Same below in `defaultSetEntryOptions`.

m-cat commented on src/skydb.js:
```diff
@@ -0,0 +1,245 @@
+ */
+
+/**
+ * Skydb v1
+ * Sets a JSON object at the registry entry corresponding to the publicKey and dataKey.
```
```suggestion
 * Sets a JSON object at the registry entry corresponding to the privateKey and dataKey using SkyDB V1.
```

parajbs replied:
I integrated it extrat so that developers and I have a better overview of which function belongs to which skydb version.

At the end, just before it's released, when we're done, I can take everything out and see how big it gets, then maybe split it into two files.

m-cat replied:
Even if it's not big, I would split it out into two separate files for the improved organization.

m-cat replied:
The docs here mention `publicKey` when they should say `privateKey`.

m-cat commented on src/skydb.js:
```diff
@@ -0,0 +1,245 @@
+ */
+
+/**
+ * Skydb v2
+ * Sets a JSON object at the registry entry corresponding to the publicKey and
+ * dataKey.
```
```suggestion
 * Sets a JSON object at the registry entry corresponding to the privateKey and
 * dataKey using SkyDB V2.
```

parajbs replied:
I integrated it extrat so that developers and I have a better overview of which function belongs to which skydb version.

At the end, just before it's released, when we're done, I can take everything out and see how big it gets, then maybe split it into two files.

m-cat commented on scripts/skydb_v2.js:
```diff
@@ -0,0 +1,200 @@
+/**
+ * Demo script for test all funktions from skydb v2.
+ *
+ * Example for Skydb v2 usage: node scripts/skydb.js dataKey "Seed Test Key" ./testdata/data.json
```
```suggestion
 * Example for Skydb v2 usage: node scripts/skydb_v2.js dataKey "Seed Test Key" ./testdata/data.json
```

m-cat commented on scripts/skydb_v2.js:
```diff
@@ -0,0 +1,200 @@
+ *
+ * Example for Skydb v2 usage: node scripts/skydb.js dataKey "Seed Test Key" ./testdata/data.json
+ *
+ * Example with default data : node scripts/skydb.js dataKey  "Seed Test Key"
```
```suggestion
 * Example with default data : node scripts/skydb_v2.js dataKey  "Seed Test Key"
```

parajbs replied:
Sorry, I was too quick there. I just fixed it in the new version

m-cat reviewed (changes requested):
This is great! There are some things that I'd like to have changed, but overall: really good job!

parajbs commented:
Hallo **m-cat,**
I just created this feature request below at "SkynetLabs/skynet-js", it will make everything much easier with the integration!
 look at it.

##################


hello **skynet-js developer,**

I'm trying to extend your repo "SkynetLabs/nodejs-skynet" with functions of "SkynetLabs/skynet-js" and make them compatible.

For the integration of the functions in nodejs-skynet, it would be very helpful if I could access the original variables and functions.
It would also be easier to handle if there are changes in "skynet-js" in the future.

Here is my idea, so that you have access to all original variables and also original functions, which are exported from "skynet-js".

To do this, I inserted a few lines in the "package.json" of "SkynetLabs/skynet-js" in the "exports" area, and I was able to convert any original variable from "skynet-js" to "nodejs-skynet" to integrate with "require".

**1. "exports" before:**

  "exports": {
    ".": {
      "import": "./dist/mjs/index.js",
      "require": "./dist/cjs/index.js"
    }

**2. "exports" my new:**

  "exports": {
    ".": {
      "import": "./dist/mjs/index.js",
      "require": "./dist/cjs/index.js"
    },
    "./\*": "./dist/cjs/\*.js",
    "./mysky/\*": "./dist/cjs/mysky/\*.js",
    "./skylink/\*": "./dist/cjs/skylink/\*.js",
    "./utils/\*": "./dist/cjs/utils/\*.js"
  },

It definitely works!

Would be very, very helpful, and a cool feature too.
You could then create better apps for the command line.

I would be happy if you integrate it. But I can also create a PR for the few lines, but only if you want it.

Let me know what you think of it.


Ask for Helpful feature for compatibility changes in package.json for developers

Sorry about the title but I had no other idea for the title

parajbs commented:
hello **m-cat**,

Ok I can fix the error with "prettier".
Was I too fast? Heck, I thought my build script worked. I need to fix my script.

I can make all the changes, but I'd rather wait to see what comes of the requests to make the change.

I would change my strategy and try to import the variables and functions directly from "skynet-js" and not port them.

**From my point of view it is more sustainable and stable for "nodejs-skynet".**

parajbs commented:
hello m-cat

Just pushed a new, improved version to github.

Cool. green heel again.

parajbs commented:
hello m-cat

I can't explain all the errors in the check test, because my "yarn test" runs cleanly.
It can also happen to me that the "yarn test" didn't work, but 5 minutes later the "yarn test" worked 6 times.
So it was on the internet, because it only works if everything is ok online and the connections are ok.

Make the changes quickly and send them up and then hope that everything is ok online again.

m-cat reviewed (changes requested):
Looks really good! I think the tests might be failing because we are running 9 jobs in parallel, and each job is accessing the same data - causing some concurrency issues. It can be solved by using a random seed, i.e. using the `genKeyPairAndSeed` function. After that, I'm ready to approve!

parajbs commented:
hello -m-cat

It was not clear to me with the 9 jobs that these mean the errors.
Did I understand correctly that I should use "genKeyPairAndSeed" instead of "genKeyPairFromSeed" in the tests.
I'll see if it works and send it back corrected.
Unless you mean something else.
I'll try that.

m-cat reviewed (dismissed):
Looks good! Needs one more review; pinged some other team members.

ro-tex reviewed (changes requested):
Looks good overall.

There is one edge case which needs to be covered (the one `REQ` I left) but that's the only blocker.

There are a number of naming inconsistencies which we generally avoid at all costs, e.g. `SeedKey` being PascalCase, which would identify it as a class or `path_dataJson` which mixes snake_case and camelCase. We typically want all functions and variables to use camelCase, all classes - PascalCase, all constants SCREAMING_SNAKE_CASE (I love this name, btw).
That said, these are minor issues that we can address in a F/U.

m-cat commented on src/skydb.js:
```diff
@@ -0,0 +1,100 @@
+  // Fetch the current value to find out the revision.
+  const signedEntry = await client.registry.getEntry(publicKey, dataKey, DEFAULT_GET_ENTRY_OPTIONS);
+
+  const revision = getNextRevisionFromEntry(signedEntry.entry);
```
This code looks like a faithful reproduction of the essential `skynet-js` logic, so any confusion originates from that repo. My apologies - the docstring could be better, as well as maybe how stuff is named. But let's not bother the contributor with this here.

> our approach should be:
    1. Get entry.
    2. Modify entry + increment revision.
    3. Save entry.

This is essentially what it does. I might have just named things in a confusing way.

parajbs commented:
Hello **ro-tex,**

Thank you just that you had time so quickly.

Sorry, because of a problem with "PascalCase" / "snake_case" and "camelCase".

I hope it's true now.
I did my best, but it may be that I overlooked something.

**I also executed your changes.**
Furthermore, I ported an important "formatSkylink" function of "skynet-js" and used it in my code.

If there is still something to complain about, report and I'll change it.

I am currently working on the integration of other functions (e.g. pinSkylink, etc.). I will put them in a new PR later.

ro-tex reviewed (dismissed):
Nice! Thank you for doing that.

LGTM!

parajbs commented:
hello **ro-tex,**

I have one more question, and what do you mean by:
"address in a F/U."
I didn't understand "F/U".

ro-tex commented:
> I have one more question, and what do you mean by: "address in a F/U." I didn't understand "F/U".

Follow up. A future issue or PR which focuses on that particular thing. Internal abbreviation, I will be more careful with its use in the future.

m-cat reviewed (dismissed):
Looks even better now! Great job.

m-cat commented:
@parajbs Oh shucks. You need your commits to be signed, otherwise github will not let us merge the pull request.

https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits

Once you've set that up, you will have to sign all the commits on this branch:

https://stackoverflow.com/a/67463502/6085242 (If that doesn't work, try https://superuser.com/a/1123928/664691)

and then you will have to force-push this branch.

Let me know if you need any help! We're so close!

parajbs commented:
Hello M-Cat,

There are problems, I first command:

git rebase --exec 'git commit --amend --no-edit -n -S' -i origin/HEAD

But didn't work because I have Windows.
Then these three commands:

1.  git rebase -i --root

**error message:**
Auto-merging src/upload.js
CONFLICT (content): Merge conflict in src/upload.js
error: could not apply 2c28e87... Added dryRun option to uploads
hint: Resolve all conflicts manually, mark them as resolved with
hint: "git add/rm <conflicted_files>", then run "git rebase --continue".
hint: You can instead skip this commit: run "git rebase --skip".
hint: To abort and get back to the state before "git rebase", run "git rebase --abort".
Could not apply 2c28e87... Added dryRun option to uploads

2.  git commit -S --amend --no-edit && git rebase --continue

**error message:**
error: Committing is not possible because you have unmerged files.
hint: Fix them up in the work tree, and then use 'git add/rm <file>'
hint: as appropriate to mark resolution and make a commit.
fatal: Exiting because of an unresolved conflict.
U       src/upload.js

3.  git add src/upload.js

Should I take the third step too?
Because I didn't change the "SRC/Upload.Js".
I'm not sure if I should do it?

Please help

parajbs commented:
Or do I have to update at GitHub in my repo with the master of "Skynetlabs/nodejs-skynet"?

parajbs commented:
hello **m-cat and MSevey,**

how should I proceed?

But you can also continue to try to solve the problems here.
Or will you do the rest?

Let me know.

m-cat commented:
@parajbs Normally we require all commits to be signed, but we made an exception in your case and merged this in.

Thank you for all of your hard work. 👏

parajbs commented:
Hello m-cat

I just thought about it again and I think that updating the RP shouldn't hurt.

If signing still fails. We have to see if I / you or Google can think of anything else.

i test it.
